### PR TITLE
Support `arrow:schema` in Parquet writer to faithfully roundtrip `duration` types with Arrow

### LINKED ISSUE #15847: [FEA] Support `arrow:Schema` in Parquet writer for faithful roundtrip with Arrow via Parquet
Support writing `arrow:Schema` to parquet file footer to allow faithful interop with Arrow via Parquet. The writer's side may be much more complicated and involve supporting all types, and computing and writing dictionary ids.

Related to: #15617

```[tasklist]
### Tasks
- [ ] #15875
- [ ] #15901
```

### PR DESCRIPTION
**Pull request by mhaseeb123**
## Description

Closes #15847

This PR adds the support to construct and write base64-encoded serialized `arrow:schema`-type IPC message to parquet file footer to allow faithfully roundtrip with Arrow via Parquet for `duration` type.

### Answered
- [x] Only construct and write `arrow:schema` if  asked by the user via `store_schema` argument (cudf) or `write_arrow_schema` (libcudf). i.e. Default these variables to `false` otherwise.
- [x] The internal/libcudf variable name for `store_schema` can stay `write_arrow_schema` and it should be fine. This has been done to disambiguate which schema (arrow or parquet) we are talking about.
- [x] Separate PR: `int96_timestamps` cannot be deprecated/removed in cuDF as Spark is actively using it. #15901 
- [x] cuDF Parquet writer supports `decimal32` and `decimal64` [fixed types](https://github.com/rapidsai/cudf/blob/branch-24.08/cpp/src/io/parquet/writer_impl.cu#L561). These are not directly supported by Arrow so we will [convert](https://github.com/rapidsai/cudf/blob/branch-24.08/cpp/src/interop/to_arrow.cu#L155) `decimal32/decimal64` columns to `decimal128`.
- [x] `is_col_nullable()` function moved to `writer_impl_helpers.cpp` along with some other helper functions.
- [x] A common `convert_data_to_decimal128` can be separated out and used in `writer_impl.cu` and `to_arrow.cu`. Tracking in a separate issue. #16194 

CC @vuule @etseidl @nvdbaranec @GregoryKimball @galipremsagar for vis. 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/io/parquet/arrow_schema_writer.cpp [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+    if constexpr (std::is_same_v<T, cudf::list_view>) {
+      children.emplace_back(make_arrow_schema_fields(
+        fbb, col->children[1], col_meta.child(1), write_mode, utc_timestamps));
+      type_type    = flatbuf::Type_List;
+      field_offset = flatbuf::CreateList(fbb).Union();
+    }
+
+    // Traverse the struct in DFS manner and process children fields.
+    else if constexpr (std::is_same_v<T, cudf::struct_view>) {
+      std::transform(thrust::make_counting_iterator(0UL),
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/arrow_schema_writer.cpp:263]: This cannot be converted to `zip_iterator` as counter is needed for `col_meta.child(idx)`

File: python/cudf/cudf/_lib/parquet.pyx [hunk 2]
```diff
@@ -621,7 +627,8 @@ cdef class ParquetWriter:
                   int max_page_size_bytes=524288,
                   int max_page_size_rows=20000,
                   int max_dictionary_size=1048576,
-                  bool use_dictionary=True):
+                  bool use_dictionary=True,
+                  bool store_schema=True):
```
Comment (mhaseeb123) [inline python/cudf/cudf/_lib/parquet.pyx:631]: Should we change the internal arg names `write_arrow_schema` to `store_schema` (used in pyarrow) or let it be?

File: python/cudf/cudf/_lib/parquet.pyx [hunk 3]
```diff
@@ -642,6 +649,7 @@ cdef class ParquetWriter:
             if use_dictionary
             else cudf_io_types.dictionary_policy.NEVER
         )
+        self.write_arrow_schema = store_schema
```
Comment (mhaseeb123) [inline python/cudf/cudf/_lib/parquet.pyx:652]: Should we change the internal arg name write_arrow_schema to store_schema (used in pyarrow) or let it be?

File: python/cudf/cudf/tests/test_parquet.py [hunk 4]
```diff
@@ -3407,3 +3410,143 @@ def test_parquet_reader_roundtrip_structs_with_arrow_schema():
...
+                                "Duration": datetime.timedelta(seconds=1),
+                            }
+                        ],
+                    }
+                }
+            }
+        ],
+    ],
+)
+def test_parquet_writer_roundtrip_structs_with_arrow_schema(tmpdir, data):
```
Comment (mhaseeb123) [inline python/cudf/cudf/tests/test_parquet.py:3529]: Still need to add tests using parquet readers and writers (both simple and chunked)

File: cpp/src/io/parquet/arrow_schema_writer.cpp [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+      type_type = flatbuf::Type_Decimal;
+      field_offset =
+        flatbuf::CreateDecimal(fbb, col_meta.get_decimal_precision(), col->type().scale(), 128)
+          .Union();
+    }
+    // cuDF-PQ writer supports ``decimal32`` and ``decimal64`` types, not directly supported by
+    // Arrow without explicit conversion. See more:
+    // https://github.com/rapidsai/cudf/blob/branch-24.08/cpp/src/interop/to_arrow.cu#L155.
+    else {
+      // TODO: Should we fail here or just not write arrow schema?.
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/arrow_schema_writer.cpp:256]: Should we fail here or just not write arrow schema and keep moving?

File: cpp/src/io/parquet/writer_impl.cu [hunk 6]
```diff
@@ -455,7 +465,7 @@ struct leaf_schema_fn {
     col_schema.type           = Type::INT32;
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/writer_impl.cu:465]: `timestamp_D` should correspond to `DateType` in arrow, though there is no way in cuDF to actually create `timestamp_D` columns.

File: python/cudf/cudf/_lib/parquet.pyx (hunk 3 above)
Comment (mhaseeb123) [reply python/cudf/cudf/_lib/parquet.pyx:652]: Answered in cuIO standup (Wed 5/29): let it be like this

File: cpp/src/io/parquet/writer_impl.cu [hunk 7]
```diff
@@ -604,33 +628,19 @@ struct leaf_schema_fn {
...
-}
-
 /**
  * @brief Construct schema from input columns and per-column input options
  *
  * Recursively traverses through linked_columns and corresponding metadata to construct schema tree.
  * The resulting schema tree is stored in a vector in pre-order traversal order.
  */
-std::vector<schema_tree_node> construct_schema_tree(
+std::vector<schema_tree_node> construct_parquet_schema_tree(
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/writer_impl.cu:637]: Renamed to disambiguate between `consturct_arrow_schema_***` and `construct_parquet_schema_***`

File: cpp/src/io/parquet/writer_impl.cu [hunk 8]
```diff
@@ -2339,16 +2363,21 @@ writer::impl::impl(std::vector<std::unique_ptr<data_sink>> sinks,
...
     _write_v2_headers(options.is_enabled_write_v2_headers()),
+    _write_arrow_schema(options.is_enabled_write_arrow_schema()),
     _sorting_columns(options.get_sorting_columns()),
     _column_index_truncate_length(options.get_column_index_truncate_length()),
     _kv_meta(options.get_key_value_metadata()),
     _single_write_mode(mode),
     _out_sink(std::move(sinks)),
     _compression_statistics{options.get_compression_statistics()}
 {
+  if (options.is_enabled_int96_timestamps() and options.is_enabled_write_arrow_schema()) {
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/writer_impl.cu:2378]: Warn if arrow schema and int96 timestamps are simultaneously enabled for now. int96 timestamp removal work will be in a separate near-future PR.

File: cpp/src/io/parquet/arrow_schema_writer.cpp [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+    field_offset =
+      flatbuf::CreateTimestamp(
+        fbb, flatbuf::TimeUnit_NANOSECOND, (utc_timestamps) ? fbb.CreateString("UTC") : 0)
+        .Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, cudf::duration_D>, void> operator()()
+  {
+    type_type    = flatbuf::Type_Time;
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/arrow_schema_writer.cpp:211]: duration_D is written as `TimeType` as `duration_D` is not a valid arrow type. This also allows for easy and faithful roundtripping with cudf.

File: cpp/src/io/parquet/writer_impl.cu [hunk 10]
```diff
@@ -1169,32 +1165,6 @@ void gather_fragment_statistics(device_span<statistics_chunk> frag_stats,
   stream.synchronize();
 }
 
-auto to_nvcomp_compression_type(Compression codec)
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/writer_impl.cu:1172]: Moved all these to `writer_impl_helpers.cpp`

File: cpp/src/io/parquet/writer_impl.cu [hunk 11]
```diff
@@ -1634,23 +1604,125 @@ size_t column_index_buffer_size(EncColumnChunk* ck,
...
+ * @tparam DecimalType to convert from
+ *
+ * @param column A view of the input columns
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ *
+ * @return A device vector containing the converted decimal128 data
  */
-void fill_table_meta(std::unique_ptr<table_input_metadata> const& table_meta)
+template <typename DecimalType>
+rmm::device_uvector<__int128_t> convert_data_to_decimal128(column_view const& column,
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/writer_impl.cu:1617]: Modified from a function in `to_arrow.cu`. Should it be moved to a common place and reused?

File: cpp/src/io/parquet/writer_impl.cu [hunk 12]
```diff
@@ -183,26 +190,6 @@ struct aggregate_writer_metadata {
 
 namespace {
 
-/**
- * @brief Function that translates GDF compression to parquet compression.
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/writer_impl.cu:187]: Moved to `writer_impl_helpers.cpp`

File: cpp/src/io/parquet/writer_impl.cu [hunk 13]
```diff
@@ -2309,16 +2351,22 @@ writer::impl::impl(std::vector<std::unique_ptr<data_sink>> sinks,
     _dict_policy(options.get_dictionary_policy()),
     _max_dictionary_size(options.get_max_dictionary_size()),
     _max_page_fragment_size(options.get_max_page_fragment_size()),
-    _int96_timestamps(options.is_enabled_int96_timestamps()),
+    _int96_timestamps(options.is_enabled_int96_timestamps() and
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/writer_impl.cu:2354]: Is this okay or should we leave it as is and `CUDF_FAIL` when building arrow schema if `INT96` timestamps enabled?

File: python/cudf/cudf/io/parquet.py [hunk 14]
```diff
@@ -960,6 +962,7 @@ def to_parquet(
     column_encoding=None,
     column_type_length=None,
     output_as_binary=None,
+    store_schema=False,
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply python/cudf/cudf/io/parquet.py:965]: Same reason as above. Thanks for catching docstring. Will update that

File: python/cudf/cudf/tests/test_parquet.py [hunk 15]
```diff
@@ -3407,3 +3487,183 @@ def test_parquet_reader_roundtrip_structs_with_arrow_schema():
...
+    buffer = BytesIO()
+
+    # Write to buffer with cudf
+    expected.to_parquet(buffer, store_schema=True)
+
+    # Read parquet with arrow schema
+    got = cudf.read_parquet(buffer)
+    # Convert to cudf table for an apple to apple comparison
+    expected = cudf.from_pandas(pdf)
+
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply python/cudf/cudf/tests/test_parquet.py:3501]: Thank you for catching this.

File: python/cudf/cudf/tests/test_parquet.py [hunk 16]
```diff
@@ -3407,3 +3487,183 @@ def test_parquet_reader_roundtrip_structs_with_arrow_schema():
...
+    expected = cudf.DataFrame.from_arrow(pa_expected)
+
+    # Write expected data frame to Parquet with arrow schema
+    buffer = BytesIO()
+    expected.to_parquet(buffer, store_schema=True, index=index)
+
+    # Read Parquet with pyarrow and pandas
+    pa_got = pq.read_table(buffer)
+    pd_got = pd.read_parquet(buffer)
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply python/cudf/cudf/tests/test_parquet.py:3652]: I initially did this to ensure interoperability between `to/from_pandas()`, `to/from_arrow()` and parquet but I agree it's kinda redundant. Removing.

File: python/cudf/cudf/tests/test_parquet.py [hunk 17]
```diff
@@ -2546,6 +2555,10 @@ def normalized_equals(value1, value2):
         value1 = None
     if value2 is pd.NA or value2 is pd.NaT:
         value2 = None
+    if isinstance(value1, np.datetime64):
```
Comment (mhaseeb123) [inline python/cudf/cudf/tests/test_parquet.py:2558]: To handle occasional `np.datetime64` type timestamp columns in pyarrow tables.

File: cpp/src/io/parquet/arrow_schema_writer.cpp [hunk 18]
```diff
@@ -0,0 +1,381 @@
...
+  std::enable_if_t<std::is_same_v<T, cudf::duration_ns>, void> operator()()
+  {
+    type_type    = flatbuf::Type_Duration;
+    field_offset = flatbuf::CreateDuration(fbb, flatbuf::TimeUnit_NANOSECOND).Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<cudf::is_fixed_point<T>(), void> operator()()
+  {
+    if constexpr (not std::is_same_v<T, numeric::decimal128>) {
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/arrow_schema_writer.cpp:247]: We shouldn't really hit this `if` clause. Thinking of removing it.

File: cpp/src/io/parquet/arrow_schema_writer.cpp [hunk 19]
```diff
@@ -0,0 +1,381 @@
...
+  {
+    type_type    = flatbuf::Type_Utf8View;
+    field_offset = flatbuf::CreateUtf8View(fbb).Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, cudf::timestamp_D>, void> operator()()
+  {
+    type_type = flatbuf::Type_Date;
+    // Use one of the strings: "UTC", "Etc/UTC" or "+00:00" to indicate a native UTC timestamp
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/arrow_schema_writer.cpp:162]: Erroneously copy-pasted comment. Will remove in any new commit.

File: cpp/include/cudf/io/parquet.hpp [hunk 20]
```diff
@@ -824,6 +833,21 @@ class parquet_writer_options_base {
...
+  void enable_write_arrow_schema(bool val);
+
+  /**
+   * @brief Sets column chunks file path to be set in the raw output metadata.
+   *
+   * @param file_paths Vector of Strings which indicates file path. Must be same size as number of
+   * data sinks in sink info
+   */
+  void set_column_chunks_file_paths(std::vector<std::string> file_paths);
+
```
Comment (etseidl) [inline cpp/include/cudf/io/parquet.hpp:850]:
```suggestion
```

Leftover from merge?

Comment (mhaseeb123) [reply cpp/include/cudf/io/parquet.hpp:850]: Thanks for catching this, Ed! Yep, leftover from merge.

File: cpp/src/io/parquet/arrow_schema_writer.cpp (hunk 9 above)
Comment (vuule) [reply cpp/src/io/parquet/arrow_schema_writer.cpp:212]: please leave a comment

File: cpp/src/io/parquet/arrow_schema_writer.cpp (hunk 1 above)
Comment (vuule) [reply cpp/src/io/parquet/arrow_schema_writer.cpp:279]: I think thrust::tabulate would work here, it just requires `children.resize(col->children.size());` beforehand.

**Review by vuule — COMMENTED** [review]
partial review

File: cpp/src/io/parquet/parquet.hpp [hunk 21]
```diff
@@ -57,6 +57,15 @@ struct TimeUnit {
   Type type;
 };
 
+struct DateUnit {
```
Comment (etseidl) [inline cpp/src/io/parquet/parquet.hpp:60]: Why are we inventing a non-existent parquet logical type? I can see this leading to much confusion in the future...at a minimum this should be documented. I see that you set the logical type to `DATE` ultimately and don't persist the `DateUnit`, but this should be made clear somewhere. Since there's already a precedent for adding to cudf specific fields to `SchemaElement`, perhaps this extra date info could go there instead?

File: cpp/src/io/parquet/writer_impl_helpers.hpp [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+void fill_table_meta(std::unique_ptr<table_input_metadata> const& table_meta);
+
+/**
+ * @brief Compute size (in bytes) of the data stored in the given column.
+ *
+ * @param column The input column
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return The data size of the input
+ */
+size_t column_size(column_view const& column, rmm::cuda_stream_view stream);
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/writer_impl_helpers.hpp:61]: > surprised this does not exist outside of cuIO

Yeah I was thinking that too when moving it.

File: cpp/src/io/parquet/arrow_schema_writer.cpp (hunk 1 above)
Comment (mhaseeb123) [reply cpp/src/io/parquet/arrow_schema_writer.cpp:279]: I think `thrust::tabulate` over `children` would be similar to `for_each` but we need the counter (idx) to access `col_meta.child(..)` unless we make it static in the lambda function which may get messy.

File: cpp/src/io/parquet/writer_impl.cu [hunk 23]
```diff
@@ -521,62 +464,80 @@ struct leaf_schema_fn {
...
-    col_schema.logical_type   = LogicalType{TimeType{timestamp_is_utc, TimeUnit::MILLIS}};
+    col_schema.type = (write_arrow_schema) ? Type::INT64 : Type::INT32;
+    col_schema.stats_dtype =
+      (write_arrow_schema) ? statistics_dtype::dtype_int64 : statistics_dtype::dtype_int32;
+    // only write as time32 logical type if not writing arrow schema
+    if (not write_arrow_schema) {
+      col_schema.converted_type = ConvertedType::TIME_MILLIS;
+      col_schema.logical_type   = LogicalType{TimeType{timestamp_is_utc, TimeUnit::MILLIS}};
+      col_schema.ts_scale       = 1000;
+    }
```
Comment (etseidl) [inline cpp/src/io/parquet/writer_impl.cu:475]:
```suggestion
    // if writing arrow schema, no logical type nor converted type is necessary
    if (write_arrow_schema) {
      col_schema.type        = Type::INT64;
      col_schema.stats_dtype = statistics_dtype::dtype_int64;
    } else {
      // parquet TIME_MILLIS annotates INT32
      col_schema.type           = Type::INT32;
      col_schema.stats_dtype    = statistics_dtype::dtype_int32;
      col_schema.converted_type = ConvertedType::TIME_MILLIS;
      col_schema.logical_type   = LogicalType{TimeType{timestamp_is_utc, TimeUnit::MILLIS}};
      col_schema.ts_scale       = 1000;
    }
```
Not sure if it's a little cleaner looking getting rid of the multiple ternary operators (same comment below for `duration_ms`).

Comment (mhaseeb123) [reply cpp/src/io/parquet/writer_impl.cu:475]: I don't mind either or. I guess we can ask folks for their preference here. @vuule?

Comment (vuule) [reply cpp/src/io/parquet/writer_impl.cu:475]: not a strong preference, but the single if is easier to follow

File: cpp/src/io/parquet/parquet.hpp (hunk 21 above)
Comment (mhaseeb123) [reply cpp/src/io/parquet/parquet.hpp:60]: I agree. I initially added this for some semantic consistency with arrow which has both Date32 and Date64 types but eventually both are converted/hardcoded to Parquet's Date (aka Date32) type regardless so while this may make sense in cudf types (if and when we are supporting Date64), it doesn't make sense in Parquet. Reverting the change.

File: cpp/src/io/parquet/writer_impl_helpers.hpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/writer_impl_helpers.hpp]: As did I when I wrote it 😂

**Comment by mhaseeb123** [comment]
The failing tests are due to some internal GH error and not related to PR.

File: cpp/src/io/parquet/writer_impl_helpers.hpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/writer_impl_helpers.hpp]: A good question is: Should it be a cudf-wide available utility for future developers?

[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/writer_impl_helpers.hpp]: The conservative in me says wait until it's needed elsewhere, cudf has gotten along just fine so far without it. 😉

[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/writer_impl_helpers.hpp]: "All right then, keep your secrets"

**Review by vuule — COMMENTED** [review]
few more comments, still haven't finished a full pass

File: cpp/src/io/parquet/arrow_schema_writer.cpp [hunk 24]
```diff
@@ -0,0 +1,377 @@
...
+ * @brief Functor to convert cudf column metadata to arrow schema field metadata
+ */
+struct dispatch_to_flatbuf {
+  FlatBufferBuilder& fbb;
+  cudf::detail::LinkedColPtr const& col;
+  column_in_metadata const& col_meta;
+  single_write_mode const write_mode;
+  bool const utc_timestamps;
+  Offset& field_offset;
+  flatbuf::Type& type_type;
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/arrow_schema_writer.cpp:71]: I did name it as `type_type` (the same variable name in Arrow and flatbuffer code) but a better name would be `field_type_id` so renaming to that instead.

File: cpp/src/io/parquet/writer_impl_helpers.cpp [hunk 25]
```diff
@@ -0,0 +1,138 @@
...
+    return ret;
+  } else if (column.type().id() == type_id::LIST) {
+    auto const lcol = lists_column_view(column);
+    return column_size(lcol.get_sliced_child(stream), stream);
+  }
+
+  CUDF_FAIL("Unexpected compound type");
+}
+
+[[nodiscard]] bool is_col_nullable(cudf::detail::LinkedColPtr const& column,
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/writer_impl_helpers.cpp:114]: The function should technically work for input columns `column_view` instead of `cudf::detail::LinkedColPtr` as well since `column->null_count()` is available in both classes. We could make it a template as well but I don't think it's being used elsewhere yet.

[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/writer_impl_helpers.cpp:114]: In short, I don't mind it being renamed to `is_output_column_nullable` so either or is good!

[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/writer_impl_helpers.cpp:114]: honestly, it's only used for schema construction...it would probably be better to just leave it in `writer_impl.cu` or turn it into a lambda in `construct_schema_tree()`.

[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/parquet/writer_impl_helpers.cpp:114]: the reason I'm suggesting this is because AFAICT this function determines, from the input column and writer options, if the written column should be nullable. It's logic specific to the PQ writer (again, IIUC).

[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/writer_impl_helpers.cpp:114]: You are right but looking at this function as-is, it takes in an arbitrary column view and returns its nullability. The returned nullability in our case is used to set the corresponding the output column, yes, but the function can be used outside this context as well if ever needed.

[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/writer_impl_helpers.cpp:114]: All in all, I don't have a strong preference one way or the other (happy to do it if folks like) but was just putting my thoughts here.

[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/writer_impl_helpers.cpp:114]: The only thing that keeps it from being general is the bit about `write_mode == single_write_mode::NO`. Given this is used only to set the nullability in the output schema, it's probably best to use @vuule's suggestion to avoid future confusion.

[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/writer_impl_helpers.cpp:114]: Sounds good. Makes sense!

File: cpp/src/io/parquet/writer_impl_helpers.hpp [hunk 26]
```diff
@@ -0,0 +1,107 @@
...
+ *
+ * @param column The input column
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return The data size of the input
+ */
+[[nodiscard]] size_t column_size(column_view const& column, rmm::cuda_stream_view stream);
+
+/**
+ * @brief Returns ``true`` if the column is nullable or if the write mode is not
+ *        set to write the table all at once instead of chunked
```
Comment (etseidl) [inline cpp/src/io/parquet/writer_impl_helpers.hpp:83]:
```suggestion
 * @brief Indicates if the column should be marked as nullable in the output schema
```
Let's keep `@brief` brief :) The impact of `write_mode` can be handled in the extended description. I also think this description is closer to the intent of the function, as the name change suggests.

File: cpp/src/io/parquet/writer_impl_helpers.hpp [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ * This doesn't check every row, so assumes string and list columns are not fixed, even
+ * if each row is the same width.
+ * TODO: update this if FIXED_LEN_BYTE_ARRAY is ever supported for writes.
+ *
+ * @param column A view of the column
+ *
+ * @return Whether the column has a fixed size
+ */
+[[nodiscard]] bool is_col_fixed_width(column_view const& column);
```
Comment (etseidl) [inline cpp/src/io/parquet/writer_impl_helpers.hpp:107]: Since this one isn't used for arrow, can it be left in `writer_impl.cu`? 🙏 

I'm looking at addressing the TODO since FLBA writing is supported now.

Comment (mhaseeb123) [reply cpp/src/io/parquet/writer_impl_helpers.hpp:107]: Certainly but just to clarify, we have moved this and other helper functions to `writer_impl_helpers.cpp` to simplify `writer_impl.cu` a bit and not due to `arrow_schema_writer.cpp` in particular. Is there a particular reason to move it back to `writer_impl.cu` such as it needing CUDA code for FLBA?

Comment (etseidl) [reply cpp/src/io/parquet/writer_impl_helpers.hpp:107]: Ah, I was thinking it was a just-in-case kind of move. Never mind then, I'll just deal with the merge.

**Review by nvdbaranec — COMMENTED** [review]
Generally looks great.  Just a couple small things so far.

File: python/cudf/cudf/tests/test_parquet.py [hunk 28]
```diff
@@ -1617,7 +1617,13 @@ def test_parquet_writer_int96_timestamps(tmpdir, pdf, gdf):
     assert_eq(pdf, gdf)
 
     # Write out the gdf using the GPU accelerated writer with INT96 timestamps
-    gdf.to_parquet(gdf_fname.strpath, index=None, int96_timestamps=True)
+    # TODO: store_schema must be false when working with INT96 timestamps
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply python/cudf/cudf/tests/test_parquet.py:1620]: Actually `store_schema` is False by default so explicitly setting it False is kinda redundant and can be removed.

[the remark this answers is not in the document]
Comment (mhaseeb123) [reply python/cudf/cudf/tests/test_parquet.py:1620]: I don't think we need to as `CUDF_EXPECTS` in C++ will throw an exception so unless specifically handling them in Python for some reason, it would be redundant to test for this specific case

[the remark this answers is not in the document]
Comment (bdice) [reply python/cudf/cudf/tests/test_parquet.py:1620]: The Cython bindings layer must handle and propagate the exception. I recommend adding the test, as we have had issues with this in the past.

[the remark this answers is not in the document]
Comment (mhaseeb123) [reply python/cudf/cudf/tests/test_parquet.py]: Added

**Review by galipremsagar — APPROVED** [review]

File: cpp/src/io/parquet/writer_impl.cu [hunk 29]
```diff
@@ -1629,23 +1602,127 @@ size_t column_index_buffer_size(EncColumnChunk* ck,
...
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(column.num_children()),
+      std::back_inserter(converted_children),
+      [&](auto const idx) { return convert_column(column.child(idx), metadata.child(idx)); });
+
+    // Process this column view. Only convert if decimal32 and decimal64 column.
+    switch (column.type().id()) {
+      case type_id::DECIMAL32:
+        // Convert data to decimal128 type
+        d128_vectors.emplace_back(convert_data_to_decimal128<int32_t>(column, stream));
```
Comment (vuule) [inline cpp/src/io/parquet/writer_impl.cu:1676]: Merging these into a single kernel (per decimal type) would be great for performance with many decimal columns, but looks like it would not fit well into the recursive implementation.
How about a stream pool? 4-8 streams that we use in round robin order might help when we have to convert may decimal columns.

**Review by vuule — COMMENTED** [review]
great stuff!
just a few comments

File: cpp/src/io/parquet/writer_impl_helpers.hpp [hunk 30]
```diff
@@ -0,0 +1,97 @@
...
+ * @return Maximum compressed chunk size
+ */
+size_t max_compression_output_size(Compression codec, uint32_t compression_blocksize);
+
+/**
+ * @brief Fill the table metadata with default column names.
+ *
+ * @param table_meta The table metadata to fill
+ */
+void fill_table_meta(std::unique_ptr<table_input_metadata> const& table_meta);
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/writer_impl_helpers.hpp:70]: This helper function fills any missing column names in the `table_input_metadata` pointed by `writer::impl::_table_meta` at https://github.com/rapidsai/cudf/blob/branch-24.08/cpp/src/io/parquet/writer_impl.cu#L2381. 

Changing this to the suggested would add overhead of releasing the object and passing it to `fill_table_meta` by (non-const) reference and then reassigning it to the unique_ptr.

File: cpp/src/io/parquet/writer_impl.cu (hunk 29 above)
Comment (mhaseeb123) [reply cpp/src/io/parquet/writer_impl.cu:1676]: Tracking this in a separate issue #16194.

File: cpp/src/io/parquet/writer_impl_helpers.hpp (hunk 30 above)
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/io/parquet/writer_impl_helpers.hpp]: Done!

File: cpp/src/io/parquet/writer_impl_helpers.cpp [hunk 31]
```diff
@@ -0,0 +1,131 @@
+/*
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/writer_impl_helpers.cpp:1]: All code here has been moved from `writer_impl.cu` as is for the most part.

File: cpp/src/io/parquet/writer_impl_helpers.hpp [hunk 32]
```diff
@@ -0,0 +1,97 @@
+/*
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/writer_impl_helpers.hpp:1]: All code here in this file been moved from `writer_impl.cu` as is for the most part.

**Review by vuule — APPROVED** [review]

File: cpp/src/io/parquet/arrow_schema_writer.hpp [hunk 33]
```diff
@@ -0,0 +1,54 @@
...
+ * @param metadata Metadata of the columns of the table
+ * @param write_mode Flag to indicate that we are guaranteeing a single table write
+ * @param utc_timestamps Flag to indicate if timestamps are UTC
+ *
+ * @return The constructed arrow ipc message string
+ */
+std::string construct_arrow_schema_ipc_message(
+  cudf::detail::LinkedColVector const& linked_columns,
+  table_input_metadata const& metadata,
+  ::cudf::io::detail::single_write_mode const write_mode,
```
Comment (bdice) [inline cpp/src/io/parquet/arrow_schema_writer.hpp:51]: We probably don't need the leading `::` here.
```suggestion
  cudf::io::detail::single_write_mode const write_mode,
```

File: python/cudf/cudf/utils/ioutils.py [hunk 34]
```diff
@@ -322,6 +322,9 @@
 output_as_binary : set, optional, default None
     If a column name is present in the set, that column will be output as
     unannotated binary, rather than the default 'UTF-8'.
+store_schema : bool, default False
```
Comment (bdice) [inline python/cudf/cudf/utils/ioutils.py:325]: Is there a reason we think this should be False by default? It seems like faithful roundtrips with Arrow would be a benefit by default. However, it seems like enabling this feature will cast / convert some data types (e.g. "days" aren't supported, and only decimal128 is supported -- if I read the rest of this PR correctly). Are those conversions potentially lossy / do they change metadata? If so, are those conversions worth documenting?

**Review by bdice — APPROVED** [review]

File: python/cudf/cudf/utils/ioutils.py (hunk 34 above)
Comment (mhaseeb123) [reply python/cudf/cudf/utils/ioutils.py:325]: A couple of reasons why we chose to set it False by default.

One `arrow:schema` should only be used when we want to round-trip certain col types (primarily durations for now) with arrow _only_. Otherwise, cudf roundtrips with itself and arrow perfectly fine.

Second, cudf still supports int96 timestamps as Spark has actively been using it. Enabling `store_schema` breaks Sparks existing and future workflows requiring them to set this to False whenever using int96 timestamps.

Third, like you mentioned, we can't roundtrip decimal32 and decimal64 with cuDF itself with arrow::schema by default without (losslessly) converting them to decimal128.

To summarize, things work perfectly fine without arrow::schema for the most part until we need to faithfully round-trip duration with arrow only in which case it is enabled.

Comment (bdice) [reply python/cudf/cudf/utils/ioutils.py:325]: Thanks! This kind of information would be good to have in the docstrings!
